gson/src/main/java/com/google/gson/stream/JsonReader.java

```diff
@@ -890,7 +890,14 @@ public double nextDouble() throws IOException {
     if (p == PEEKED_LONG) {
       peeked = PEEKED_NONE;
       pathIndices[stackSize - 1]++;
-      return (double) peekedLong;
+      if (peekedLong >= Byte.MIN_VALUE && peekedLong <= Byte.MAX_VALUE) {
+        return (byte) peekedLong;
+      }else if (peekedLong >= Short.MIN_VALUE && peekedLong <= Short.MAX_VALUE){
+        return (short) peekedLong;
+      }else if (peekedLong >= Integer.MIN_VALUE && peekedLong <= Integer.MAX_VALUE){
+        return (int) peekedLong;
```

Comment:
I am not sure if it is worth it returning types other than `Long`. The caller can convert it themselves to `byte`, `short` or `int` if necessary.

Reply:
I think it's more rigorous to automatically adapt to numerical values